v0/destinations/hs/HSTransform-v2.js

```diff
@@ -0,0 +1,490 @@
+const get = require("get-value");
+const _ = require("lodash");
+const { MappedToDestinationKey } = require("../../../constants");
+const {
+  defaultGetRequestConfig,
+  defaultPostRequestConfig,
+  defaultRequestConfig,
+  getFieldValueFromMessage,
+  getSuccessRespEvents,
+  CustomError,
+  addExternalIdToTraits,
+  defaultBatchRequestConfig,
+  removeUndefinedAndNullValues,
+  getDestinationExternalID,
+  constructPayload,
+  isDefinedAndNotNullAndNotEmpty,
+  getDestinationExternalIDInfoForRetl
+} = require("../../util");
+const {
+  hsCommonConfigJson,
+  IDENTIFY_CRM_UPDATE_NEW_CONTACT,
+  IDENTIFY_CRM_CREATE_NEW_CONTACT,
+  MAX_BATCH_SIZE_CRM_CONTACT,
+  BATCH_IDENTIFY_CRM_CREATE_NEW_CONTACT,
+  BATCH_IDENTIFY_CRM_UPDATE_NEW_CONTACT,
+  mappingConfig,
+  ConfigCategory,
+  TRACK_CRM_ENDPOINT,
+  CRM_CREATE_CUSTOM_OBJECTS,
+  MAX_BATCH_SIZE_CRM_OBJECT,
+  BATCH_CREATE_CUSTOM_OBJECTS
+} = require("./config");
+const {
+  getTransformedJSON,
+  formatPropertyValueForIdentify,
+  searchContacts,
+  getCRMUpdatedProps,
+  getEventAndPropertiesFromConfig
+} = require("./util");
+
+/**
+ * Using New API
+ * Ref - https://developers.hubspot.com/docs/api/crm/contacts
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} propertyMap
+ * @returns
+ */
+const processIdentify = async (message, destination, propertyMap) => {
+  const { Config } = destination;
+  const traits = getFieldValueFromMessage(message, "traits");
+  const mappedToDestination = get(message, MappedToDestinationKey);
+  // if mappedToDestination is set true, then add externalId to traits
+  if (mappedToDestination) {
+    addExternalIdToTraits(message);
+  } else if (!traits || !traits.email) {
+    throw new CustomError(
+      "[HS]:: Identify without email is not supported.",
+      400
+    );
+  }
+
+  if (!Config.lookupField) {
+    throw new CustomError(
+      "lookupField is a required field in webapp config",
+      400
+    );
+  }
+
+  let contactId = getDestinationExternalID(message, "hsContactId");
+
+  if (!contactId) {
+    contactId = await searchContacts(message, destination);
+  }
+
+  const userProperties = await getTransformedJSON(
+    message,
+    hsCommonConfigJson,
+    destination,
+    propertyMap
+  );
+
+  const payload = {
+    properties: formatPropertyValueForIdentify(userProperties)
+  };
+
+  // build response
+  let endpoint;
+  const response = defaultRequestConfig();
+
+  // for rETL source support for custom objects
+  // Ref - https://developers.hubspot.com/docs/api/crm/crm-custom-objects
+  if (mappedToDestination) {
+    const { objectType } = getDestinationExternalIDInfoForRetl(message, "HS");
+    endpoint = CRM_CREATE_CUSTOM_OBJECTS.replace(":objectType", objectType);
+    response.body.JSON = removeUndefinedAndNullValues({ properties: traits });
+    response.source = "rETL";
+  } else {
+    if (contactId) {
+      // update
+      endpoint = IDENTIFY_CRM_UPDATE_NEW_CONTACT.replace(
```

Comment:
we can use templating
```
endpoint = `${IDENTIFY_CRM_CREATEE_NEW_CONTACT}:${contactId}`
```
Then we just can use one constant

Reply:
for some endpoints, the variable part can be in the middle of the string so keeping this way will maintain the structure. Secondly, by looking at the config directly we get to know the complete endpoint instead of half/in-complete/ambiguous which can be overlapping with some other